### DISCUSSION TITLE
Switch a follower node from Leader eligible node to a stand by node

### DISCUSSION THREAD

**Discussion by aley1** [dotnet/dotNext#134] [Ideas]
Hi Sakno,

From the documentation and code, I can see that we can start a node as Standby node and later promote it to a Node that can be elected as an Leader. However it looks like it's not possible to change Node back to a Standby node later when the cluster is formed and running.
Do you plan to introduce this functionality? 
It is actually important to the project I am working on.

**Comment by sakno** [comment]
Hey @aley1 , it's correct, you can't change the node back to **standby** state, at least manually. In some circumstances, this can be done by framework automatically. This state is an extension to original Raft that serves to some special purposes:
* In some cases, you may want to have read-only replicas in your cluster. Those replicas will never be elected as a leader, but can respond to read requests (linearizability is still respected). For instance, if you want to distribute read/write workloads across the cluster. In that case, you can convert standby node to regular node manually by calling _TurnIntoRegularNode_. This is safe operation of all active clients connected to that node.
* _Catch up_ phase when a new node is introduced to the cluster. A new node always started as standby node. It waits for incoming replication from the leader and when its local WAL close to what the leader has, the node converts to regular node automatically (`UnfreezeAsync` private method). This is also safe because normally a new node should not be visible to the potential clients (e.g., using Startup probe in K8s in combination with `IRaftCluster.Readiness` property).

Turning regular node to standby node may be dangerous for clients, especially, if you trying to convert leader to standby. Client might add uncommitted log entry without a chance to wait for a commit, and loose its write. In this case, uncommitted log entry will be eventually dropped from local WAL by a new elected leader. Client have to repeat its write request. Thus, converting to standby node can be done safely for _followers_ only. Even in that case, during the conversion the node should not be able to transit from _follower_ state to _candidate_ state. Otherwise, the exception should be thrown. However, this is doable.

According to that restrictions, is it still reasonable to add appropriate public method for you? If possible, could you please describe your use case and how the feature can be helpful?

**Reply by aley1** [reply]
Hi Sakno, 
Thanks for detailed explanation.
In my use case, I am building a custom replication protocol / log storage that is more suited to the type / amount of data I want to replicate. In other words, I am using your library for the Raft Leadership consensus feature but not the log replication (we may still use the existing persistent WAL provided but only for specific data)
Since I am handling the log replication as separately piece, I need to be able to downgrade a follower node which logs are out of sync with the leader so that it will not be elected as a leader in case the current Leader is suddenly unavailable. Also for a follower node that is out of sync we redirect the reads from clients to another node or the Leader.
If you can add the ability to change a node back to Standby mode as a public method, it will really help me in completing the above functionality. Thank you!

**Reply by sakno** [reply]
Checkout latest `develop` branch. All extensions to Raft algorithm now grouped in `DotNext.Net.Cluster.Consensus.Raft.Extensions` namespace with interfaces for each extensions. `IStandbyModeSupport` interface with methods you are interested in.

**Reply by sakno** [reply]
It will be release in 4.8.0 release. However, I can't tell right now the exact date of the release.

**Reply by aley1** [reply]
Thanks!
